adapters/interlocked_win32.c

```diff
@@ -1,4 +1,318 @@
+
+	return (int32_t)InterlockedAdd((volatile LONG*)addend, (LONG)value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int32_t, interlocked_and, volatile int32_t*, destination, int32_t, value) 
+{
+	/*Codes_SRS_INTERLOCKED_43_002 [ interlocked_and shall perform an atomic bitwise AND operation on the 32-bit integer values *destination and value and store the result in *destination.]*/
+	/*Codes_SRS_INTERLOCKED_43_033: [ interlocked_and shall return the initial value of *destination.]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_003: [ interlocked_and shall call InterlockedAnd from windows.h. ]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_004: [ interlocked_and shall return the initial value of *destination. ]*/
+
+	return (int32_t)InterlockedAnd((volatile LONG*)destination, (LONG)value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int16_t, interlocked_and_16, volatile int16_t*, destination, int16_t, value)
+{
+	/*Codes_SRS_INTERLOCKED_43_003 [ interlocked_and_16 shall perform an atomic bitwise AND operation on the 16-bit integer values *destination and value and store the result in *destination.]*/
+	/*Codes_SRS_INTERLOCKED_43_034: [ interlocked_and_16 shall return the initial value of *destination.]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_005: [ interlocked_and_16 shall call InterlockedAnd16 from windows.h. ]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_006: [ interlocked_and_16 shall return the initial value of *destination. ]*/
+
+	return (int16_t)InterlockedAnd16((volatile SHORT*)destination, (SHORT)value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile int64_t*, destination, int64_t, value)
+{
+	/*Codes_SRS_INTERLOCKED_43_004 [ interlocked_and_64 shall perform an atomic bitwise AND operation on the 64-bit integer values *destination and value and store the result in *destination.]*/
+	/*Codes_SRS_INTERLOCKED_43_035: [ interlocked_and_64 shall return the initial value of *destination.]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_007: [ interlocked_and_64 shall call InterlockedAnd64 from windows.h. ]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_008: [ interlocked_and_64 shall return the initial value of *destination. ]*/
+
+	return (int64_t)InterlockedAnd64((volatile LONG64*)destination, (LONG64)value);
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, int8_t, interlocked_and_8, volatile int8_t*, destination, int8_t, value)
+{
+	/*Codes_SRS_INTERLOCKED_43_005 [ interlocked_and_8 shall perform an atomic bitwise AND operation on the 8-bit integer values *destination and value and store the result in *destination]*/
+	/*Codes_SRS_INTERLOCKED_43_036: [ interlocked_and_8 shall return the initial value of *destination. ]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_009: [ interlocked_and_8 shall call InterlockedAnd8 from windows.h. ]*/
+	/*Codes_SRS_INTERLOCKED_WIN32_43_010: [ interlocked_and_8 shall return the initial value of *destination. ]*/
+
+	return (int8_t)InterlockedAnd8((volatile char*)destination, (char)value);
```

Comment:
>char [](start = 42, length = 4)

`CHAR`

Reply:
In the docs it seems like they are using the native C types: [https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedand8](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedand8)

---
In reply to: [427687405](https://github.com/Azure/azure-c-util/pull/43#discussion_r427687405) [](ancestors = 427687405)

Reply:
You're right, InterlockedExchange8 says `CHAR` but InterlockedAnd8 says `char`.

---
In reply to: [428107038](https://github.com/Azure/azure-c-util/pull/43#discussion_r428107038) [](ancestors = 428107038,427687405)